Islandora-CLAW/Syn
# Change Tomcat Valve to servlet Filter

**GitHub Issue**: https://github.com/Islandora-CLAW/CLAW/issues/736

Depends on: https://github.com/Islandora-CLAW/islandora/pull/84
Depends on: https://github.com/Islandora-CLAW/Crayfish-Commons/pull/15

# What does this Pull Request do?

* Converts the Tomcat Valve to a Filter
* Replaces the XML configuration with a YAML one.
* Changes some of the JWT claims to use standard ones.
* Makes the SettingsParser parse the config only once and then supply the 3 maps from the internal Config object.

# What's new?

* The servletfilter will fail to initialize if it can't access it's settings file or the settings file is invalid.
* In theory this filter should work on non-Tomcat container, but I haven't tried that.
* The claim changes are
     * `id` -> `webid`
     * `name` -> `sub`
     * `url` -> `iss`

* Does this change require documentation to be updated? Documentation has been updated
* Does this change add any new dependencies? no
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? 
* Could this change impact execution of existing code?

# How should this be tested?

There is no easy way to test this, it will require uninstalling Islandora. But this should hopefully work.

1. Make sure you have a current version of master from claw-playbook.
1. Delete your `roles/external` directory if it exists.
1. Download [this patch](https://gist.github.com/whikloj/b02dc0480d0f3e6f46c506b7bcbfb41d) to the main claw-playbook directory and apply with `git patch <patchfile>`.
1. Run `ansible-galaxy install --role-file=requirements.yml --roles-path=roles/external`
1. Now for some manual editing in `roles/external/Islandora-Devops.fcrepo-syn`
     1. `defaults/main.yml` change `fcrepo_syn_version: 0.1.0` to `fcrepo_syn_version: 0.1.1`
     1. `tasks/install.yml` change `url: https://github.com/Islandora-CLAW/Syn/releases/download/v{{ fcrepo_syn_version }}/islandora-syn-{{ fcrepo_syn_version }}-all.jar` to `url: https://github.com/whikloj/Syn/releases/download/v{{ fcrepo_syn_version }}/islandora-syn-{{ fcrepo_syn_version }}-all.jar`
1. Run `vagrant up`
1. Once the VM is up and running, log in `vagrant ssh`
1. Download and run [this script](https://gist.github.com/whikloj/5ddd3ff65dc6cebfd92ea4e2fdb9e2da).
1. When it tells you its ready for testing. Then please test.

You should be able to create a collection or image in Drupal and have it sync to Fedora.

But you should not be able to PUT/POST to fedora without a valid JWT or token. Like
```
ubuntu@claw:~$ curl -i -XPUT http://localhost:8080/fcrepo/rest/breaking
HTTP/1.1 401 Unauthorized
Server: Apache-Coyote/1.1
Content-Type: text/html;charset=utf-8
Content-Language: en
Content-Length: 1068
Date: Tue, 23 Jan 2018 16:51:21 GMT

<!DOCTYPE html><html><head><title>Apache Tomcat/8.0.32 (Ubuntu) - Error report</title><style type="text/css">H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}.line {height: 1px; background-color: #525D76; border: none;}</style> </head><body><h1>HTTP Status 401 - Token authentication failed.</h1><div class="line"></div><p><b>type</b> Status report</p><p><b>message</b> <u>Token authentication failed.</u></p><p><b>description</b> <u>This request requires HTTP authentication.</u></p><hr class="line"><h3>Apache Tomcat/8.0.32 (Ubuntu)</h3></body></html>
```

# Additional Notes:
Any additional information that you think would be helpful when reviewing this PR.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @Islandora-CLAW/committers
